_posts/2025-04-28-running-at-scale.adoc

```diff
@@ -0,0 +1,143 @@
+
+IMPORTANT: The default etcd size is 2 GB, and the recommended maximum is 8 GB. Exceeding the maximum limit can make the Kubernetes cluster unstable and unresponsive.
+
+image::/assets/img/running-at-scale/etcd.png[The metrics showing the progression of the etcd storage growth]
+
+Figure 1: Metrics with the progression of the etcd storage growth that leads to cluster instability.
+
+=== The size of the objects matters
+
+Not only the overall number, but also the size of the objects stored in etcd is a critical factor that can significantly impact its performance and stability. Each object stored in etcd consumes space, and as the number of objects increases, the overall size of etcd grows. The larger the object is, the more space it takes in etcd. For example, etcd can be overloaded with just a few thousand of relatively big Kubernetes objects.
+
+IMPORTANT: Even though the data stored in a `ConfigMap` cannot exceed 1 MiB by design, a few thousand of relatively big `ConfigMap` objects can overload etcd storage.
+
+In the context of Eclipse Che, the operator has a feature that merges the cluster's Certificate Authorities (CAs) bundle into a `ConfigMap` and provisions it for each user namespace, resulting in additional etcd usage, given that there is a large number of TLS certificates in the cluster.
+
+In order to disable mounting of the CA bundle using the `ConfigMap` under the path `/etc/pki/ca-trust/extracted/pem`, configure the `CheCluster` Custom Resource by setting `disableWorkspaceCaBundleMount` property to `true`. Certificates will be mounted under the path `/public-certs` in this case:
+
+```yaml
+spec:
+  devEnvironments:
+    trustedCerts:
+      disableWorkspaceCaBundleMount: true
+```
+
+TIP: More details about importing untrusted TLS certificates can be found in the link:https://eclipse.dev/che/docs/stable/administration-guide/importing-untrusted-tls-certificates/[official documentation].
+
+=== DevWorkspace objects
+
+For large Kubernetes deployments, particularly those involving a high number of custom resources such as `DevWorkspace` objects, which represent Cloud Development Environments, etcd can become a significant performance bottleneck. 
+
+IMPORTANT: Based on the load testing 6,000 of `DevWorkspace` objects consume around 2.5GB of the etcd storage.
+
+Starting from link:https://github.com/devfile/devworkspace-operator[DevWorkspace Operator] 0.34.0 version, it is possible to configure the pruner that would automatically clean up the `DevWorkspace` objects that are not in use for a certain period of time. This can be done by configuring the `DevWorkspaceOperatorConfig` object accordingly: 
```

Comment:
```suggestion
Starting from link:https://github.com/devfile/devworkspace-operator[DevWorkspace Operator] version 0.34.0, it is possible to configure a pruner that would automatically clean up `DevWorkspace` objects that were not in use for a certain period of time. This can be done by configuring the `DevWorkspaceOperatorConfig` object accordingly: 
```